Services/WatchlistService.swift

```diff
@@ -6,76 +6,129 @@
 //
 
 import UIKit
+import CoreData
 
 protocol WatchlistServiceProtocol {
-    func getStatus(for id: Int)-> WatchlistStatus
-    func toggleStatus(for item: WatchlistItem) -> WatchlistStatus
-    func getWatchlistItems() -> [WatchlistItem]
+    func getStatus(for id: Int) throws -> WatchlistStatus
+    func toggleStatus(for item: WatchlistMovieConfiguration)-> WatchlistStatus
+    func getWatchlist() -> [WatchlistMovie]
 }
 
 enum WatchlistStatus {
     case added, notAdded
 }
 
-class WatchlistService: WatchlistServiceProtocol {
-    typealias Watchlist = [Int : WatchlistItem]
-    private static let watchlistKey: String = "watchlist"
+enum WatchlistServiceError: Error, ErrorViewHandleable {
+    case failedToFetchFromPersistentStore
     
-    func getStatus(for id: Int)-> WatchlistStatus {
-        let watchlist = getWatchlist()
-        
-        return watchlist[id] != nil ? .added : .notAdded
+    var errorTitle: String {
+        switch self {
+        case .failedToFetchFromPersistentStore:
+            return "Houston, we have a problem.\nClose and re-open the app."
+        }
     }
     
-    func toggleStatus(for item: WatchlistItem)-> WatchlistStatus {
-        var watchlist = getWatchlist()
-        
-        let currentStatus: WatchlistStatus = watchlist[item.id] != nil ? .added : .notAdded
-        switch currentStatus {
-        case .added:
-            watchlist.removeValue(forKey: item.id)
-            return saveWatchlist(watchlist) ? .notAdded : .added
-        case .notAdded:
-            watchlist[item.id] = item
-            return saveWatchlist(watchlist) ? .added : .notAdded
+    var errorImage: UIImage? {
+        switch self {
+        case .failedToFetchFromPersistentStore:
+            return UIImage(named: "dizzy")
         }
     }
+}
+
+class Container {
+    static let shared: NSPersistentContainer = {
+        let container = NSPersistentContainer(name: "WatchlistMovies")
+        container.viewContext.mergePolicy = NSMergeByPropertyObjectTrumpMergePolicy
+        container.loadPersistentStores {storeDescription, error in
+            if let error = error {
+                print("Unresolved error \(error)")
+            }
+        }
+        return container
+    }()
     
-    func getWatchlistItems() -> [WatchlistItem] {
-        let watchlist = getWatchlist()
+    private init() {
+    }
+}
+
+class WatchlistService: WatchlistServiceProtocol {
+    let container = Container.shared
+    var predicate: NSComparisonPredicate?
+
+    func getStatus(for id: Int) throws -> WatchlistStatus {
+        let request = WatchlistMovie.createFetchRequest()
+        predicate = NSComparisonPredicate(format: "id == %@", id as NSNumber)
+        request.predicate = predicate
         
-        var items = [WatchlistItem]()
-        for value in watchlist.values {
-            items.append(value)
+        do {
+            let movies = try container.viewContext.fetch(request)
```

Comment:
Since we are only interested in the count of results we can use. No big deal, it's just a small performance improvement.
```suggestion
            let movies = try container.viewContext.count(for: request)
```

Reply:
Great, thanks!